at_cmd asynchronous command handling

bjda opened the pull request:
Reimplements much of at_cmd to enable asynchronous handling of commands. This PR focuses on keeping `at_cmd_write` working as before while laying the groundwork for asynchronous handling.

The metadata for commands and responses is stored in message queues, while the strings are allocated on the heap if required. The handling logic is designed to accept flags that enable easy expansion of the interface for different use cases (e. g. executing a command from static memory, where the command string need not be copied). This expansion could take the form of exposing the flags (or even the whole `cmd_item` struct), and/or making convenience functions for the different use cases.

Since the changes to at_cmd.c are relatively extensive, it might make more sense to review that file as a whole, rather than as a change from the previous version.

rlubos commented on lib/at_cmd/at_cmd.c:
```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+		    resp.buf != NULL) {
+			resp.callback(resp.buf);
+		}
+		if (resp.flags & AT_RESP_ALLOC) {
```
Again, I wonder if we need this flag. Isn't `callback_worker` processing only asynchronous commands and notifications, that will always have this buffer dynamically allocated (with current implementation)?

bjda replied:
Removed

bjda replied:
No longer relevant

rlubos commented on lib/at_cmd/at_cmd.c:
```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+		if (ret.state == AT_CMD_NOTIFICATION ||
+		    (current_cmd.buf == NULL &&
+		     current_cmd.flags & AT_CMD_BUF_RESP)){
+			ret.buf = k_malloc(payload_len);
```
I understand it's needed with current apporach (responses/notifications processed in a system queue), but I just wanted to express my concern, that his may introduce unreliability in notification forwarding in applications that make heavy use of a heap. This might impact modules like Link Controller.

bjda replied:
Notifications no longer allocate memory

rlubos commented on lib/at_cmd/at_cmd.c:
```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+		}
 	}
+}
+
```
A lot of (unneeded) empty lines here.

bjda replied:
Fixed

rlubos commented on lib/at_cmd/at_cmd.c:
```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+
+	/* This cast is safe; we do not free cmd without AT_CMD_BUF_CMD */
+	command.cmd = (char *)cmd;
+
```
Unneeded empty line.

bjda replied:
I almost always separate a block with empty lines when using a comment, to make it obvious what the comment refers to. I agree it is pretty obvious here, though. Removed.

rlubos commented on lib/at_cmd/at_cmd.c:
```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+
+		/* Do not load a new command if already loaded or none queued */
+		if (current_cmd.cmd != NULL ||
+		k_msgq_get(&commands, &current_cmd, K_NO_WAIT) != 0) {
```
Alignment seems off here.

bjda replied:
Fixed

rlubos commented on lib/at_cmd/at_cmd.c:
```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 	}
-
-	return ret.code;
+	memcpy(temp, cmd, strlen(cmd) + 1);
```
We deal with a string, you can use `strcpy`.

bjda replied:
Done

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
 
-	k_mem_slab_free(&rsp_work_items, (void **)&data);
+/* Distributes an AT response to the appropriate queue */
+static void distribute_resp(const struct cmd_item *cmd, struct resp_item *resp)
```
Nit: I would call this function `dispatch_resp` instead.

bjda replied:
The function no longer exists

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+
+	LOG_DBG("Sending command %s", log_strdup(cmd));
+
+	bytes_sent = send(common_socket_fd, cmd, bytes_to_send, 0);
```
Sending a null-terminated String should work.

bjda replied:
I am not sure I follow this one. Could you elaborate on what is suggested?

bjda replied:
Handled offline

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -111,23 +134,114 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+	if (bytes_sent != bytes_to_send) {
+		LOG_ERR("Bytes sent (%d) was not the "
+			"same as expected (%d)",
+			bytes_sent, bytes_to_send);
```
Missing return.
But I think you can make this simpler. Just print and return the same error for both.

bjda replied:
The lack of return here is inherited from the current implementation (64bdc76384a31cc60d10aeb9a37795cdd2175ed2) and, I think, intentional. If a partial AT command is in fact sent, at_cmd has to wait for the (error) response. It seems like an unnecessary check, if `send` either fails or sends the correct length. @evenl  , do you know if it is still needed?

bjda replied:
Handled offline

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 
-static inline int at_write(const char *const cmd, enum at_cmd_state *state)
+/* Allocate and copy cmd to a new buffer */
+static char *buffer_cmd(const char *cmd)
```
Do you really need a function for this? You are calling it only once.

bjda replied:
Not at all. Integrated it into `at_cmd_write_with_callback`

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -171,124 +290,130 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
-	response_buf_len = buf_len;
+	if (cmd == NULL) {
+		LOG_ERR("cmd is NULL");
+		*state = AT_CMD_ERROR_QUEUE;
```
Do you really need to set an error state? is the error code not enough?

bjda replied:
The error state is part of the current API, but was expanded to clarify what stage the error relates to. To clarify, are you suggesting I remove/rework the state part from the API, remove the more specific error states, or forego setting the state for this error?

bjda replied:
Handled offline

metc commented:
cc @evenl

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=2264) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/nrfconnect/sdk-nrf?pullRequest=2264) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/nrfconnect/sdk-nrf?pullRequest=2264) it.</sub>

bjda commented on lib/at_cmd/at_cmd.c:
```diff
@@ -22,34 +22,57 @@ LOG_MODULE_REGISTER(at_cmd, CONFIG_AT_CMD_LOG_LEVEL);
-				CONFIG_AT_CMD_THREAD_STACK_SIZE);
+/* Flags describing an AT command request */
+enum at_cmd_flags {
+	AT_CMD_BUF_CMD = 1 << 0, /* Command is buffered by at_cmd */
```
Exactly. The flags are there to keep the current API working and enable easy expansion of the interface. This table was used in the design process:
![image](https://user-images.githubusercontent.com/51139291/82780644-2ac3b100-9e58-11ea-9933-6c72e25461d2.png)
There are 9 cases there, and so they cannot be differentiated without at least 4 bits (we get 1 free from checking whether `buf` is populated).

bjda replied:
AT_CMD_BUF_RESP is now removed, since responses are not buffered

bjda commented:
@rlubos I changed the implementation to call the callbacks directly from the socket thread, and added a check (assert) to eliminate the possible deadlock

rlubos commented:
@bjda Did you sync up with the `lwm2m_carrier` team, I've mentioned that calling the callback directly from the socket thread will require some involvement on their side, i.e. switching to the async `at_cmd_write_with_callback` in the notification handler:

>  But I understand this would require some involvement and "good will" from current users of this module (if I recall correctly there was a problem in lwm2m_carrier).

bjda commented:
@rlubos I have not, who would be the main contact for that?

rlubos commented:
@metc?

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -22,34 +22,44 @@ LOG_MODULE_REGISTER(at_cmd, CONFIG_AT_CMD_LOG_LEVEL);
+/* Metadata for a queued AT command */
+struct cmd_item  {
+	char *cmd;			/* Pointer to 0-terminated command */
+	char *buf;			/* Pointer to response buffer */
```
You could maybe rename `buf` and `buf_len` to `resp` and `resp_len`, and group them in the struct?
I assume the response is also a null-terminated String?

bjda replied:
Done. The grouping is to save memory

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,125 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 			close(common_socket_fd);
 			return;
-		} else if (item->data[bytes_read - 1] != '\0') {
+		} else if (buf[bytes_read - 1] != '\0') {
```
What if `bytes_read` is 0?

metc replied:
@evenl Can we read a 0-length from an AT socket?
Here we are using the common shared AT socket. If a user closes this socket, what will the return value be? Should we first check for `errno` or in case of error is the return value always -1?

evenl replied:
The errno is not valid without a -1 return value, otherwise you don't know which call left the errno. I haven't checked with the code yet if we can get a 0 here. I can try to do that later today

bjda replied:
As discussed offline, this is not a regression. However, if there is any real chance of this happening, I can rewrite the error handling part to reflect that (probably adding `bytes_read == 0` as a separate case).

evenl replied:
I have never seen it, and it doesn't make sense. But the AT interface in BSDLib allow for packages containing 0 bytes, it's not considered an error condition by BSDlib. So you can, in theory receive 0 bytes from the AT socket.

bjda replied:
Added handling for this case

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,125 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+				ret.code  = -EMSGSIZE;
 				goto next;
 			}
+			strcpy(current_cmd.buf, buf);
```
Use `strncpy` to be safe here? maybe the `payload_len` calculation can be wrong?

bjda replied:
Moved to memcpy as discussed offline

metc commented on lib/at_cmd/at_cmd.c:
```diff
@@ -109,25 +119,82 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+			if (current_cmd.flags & AT_CMD_SYNC) {
+				k_msgq_put(&response_sync, &resp, K_FOREVER);
+			}
+			complete_cmd();
```
`current_cmd_mutex` is already locked, isnt'it ?

metc commented on include/modem/at_cmd.h:
```diff
@@ -86,8 +84,7 @@ int at_cmd_init(void);
  */
 int at_cmd_write_with_callback(const char *const cmd,
-					  at_cmd_handler_t  handler,
-					  enum at_cmd_state *state);
```
The async API is not used yet but for a future version I would like to have a symmetric API. The `at_cmd_handler_t ` callback should be changed to a new one so the user could receive an error code and the result of the AT command response (state).
Would be nice to have this new API in the next version.

bjda replied:
Well noted

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
-			goto next;
+		/* Call the relevant callback, if any */
+		if (ret.state == AT_CMD_NOTIFICATION &&
+		    notification_handler != NULL) {
```
Are you sure you want to run this in the same thread and not put it on the work queue? This may end up in a deadlock of the notification handler is sending another AT command with at_cmd_write();

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+		/* Call the relevant callback, if any */
+		if (ret.state == AT_CMD_NOTIFICATION &&
+		    notification_handler != NULL) {
+			notification_handler(buf);
```
Are you sure you want to run this in the same thread and not put it on the work queue? This may end up in a deadlock of the notification handler is sending another AT command with at_cmd_write();

bjda replied:
It is at least preferred. It implements https://github.com/nrfconnect/sdk-nrf/pull/2264/files/c3e874cda311323d5ca7f825bab95f45ba063b6f#pullrequestreview-412504093 to save significant memory and avoid needing the heap.
The deadlock potential is documented and crashes out at at_cmd:319 just in case:
```c
	__ASSERT(k_current_get() != socket_tid,
		 "at_cmd deadlock: socket thread blocking self\n");
```

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+		    notification_handler != NULL) {
+			notification_handler(buf);
+		} else if (current_cmd.callback != NULL) {
+			current_cmd.callback(buf);
```
Are you sure you want to run this in the same thread and not put it on the work queue? This may end up in a deadlock of the notification handler is sending another AT command with at_cmd_write();

bjda replied:
See my answer on 261

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -109,25 +119,82 @@ static int get_return_code(char *buf, struct return_state_object *ret)
+/* Clear the current command safely */
+static void complete_cmd(void)
+{
+	k_mutex_lock(&current_cmd_mutex, K_FOREVER);
```
Isn't this mutext already locked at this point when comming from line 186.

bjda replied:
Yes, it is doubly locked in that case

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
+	}
 
-	return return_code;
+	load_cmd_and_write();
```
Doesn't this mean that you will always process the queue with exactly one command?

bjda replied:
I understand the confusion; `load_cmd_and_write` does not necessarily load a cmd and write it. As the description hints, it only does that "if appropriate". That is, only if no command is currently being processed. The logic can be seen in the if statement on line 168.

bjda replied:
In effect, that function is a synchronization mechanism that avoids having two separate threads for listening to the queue and the socket.

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -136,24 +203,25 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
-		k_mem_slab_alloc(&rsp_work_items, (void **)&item, K_FOREVER);
-		LOG_DBG("Allocation done");
+		LOG_DBG("Writing any pending command");
+		load_cmd_and_write();
```
I a bit confused why you need this, the queue is already processed when you issue the command. I can not see any way to insert commands in the queue without them beeing processed righ away.

bjda replied:
My answer on 307 should clarify this :)

evenl commented on lib/at_cmd/at_cmd.c:
```diff
@@ -162,133 +230,126 @@ static void socket_thread_fn(void *arg1, void *arg2, void *arg3)
 
-	int return_code = at_write(cmd, state);
+	LOG_DBG("Awaiting response for %s", log_strdup(cmd));
+	k_msgq_get(&response_sync, &ret, K_FOREVER);
```
If for instance a low priority thread issues a command and the goes to sleep while waiting for and response on the message queue. Then a high level thread also issue a command, and then wait for the message queue. The at_cmd service thread receive the answer to the low level thread and put it on the message queue, which of the low priority and high priority threads are then woken and receives the first answer?

rlubos replied:
That's a good point, I think we should keep the semaphore/mutex at the function entry for synchronous operation. For async it does not matter since the request is bound with the callback anyway.

bjda replied:
I have added a mutex to protect the hazardous region

bjda commented:
Rebased and adressed new comments. @evenl and @rlubos, could you revisit?
